core/ledger/kvledger/kv_ledger_test.go

```diff
@@ -567,6 +573,363 @@ func TestLedgerWithCouchDbEnabledWithBinaryAndJSONData(t *testing.T) {
+	testVerifyPvtData(t, lgr1, blockNumAtCrash, dataAtCrash.PvtData)
+	bcInfo, err = lgr.GetBlockchainInfo()
+	assert.NoError(t, err)
+	assert.Equal(t, blockNumAtCrash, bcInfo.Height)
+
+	// we should be able to write the last block again
+	// to ensure that the pvtdataStore is not updated, we send a different pvtData for
+	// the same block such that we can retrieve the pvtData and compare.
+	expectedPvtData := dataAtCrash.PvtData
+	dataAtCrash.PvtData = make(ledger.TxPvtDataMap)
+	dataAtCrash.PvtData[0] = &ledger.TxPvtData{
+		SeqInBlock: 0,
+		WriteSet: &rwset.TxPvtReadWriteSet{
+			NsPvtRwset: []*rwset.NsPvtReadWriteSet{
+				{
+					Namespace: "ns-1",
+					CollectionPvtRwset: []*rwset.CollectionPvtReadWriteSet{
+						{
+							CollectionName: "coll-1",
+							Rwset:          []byte("pvtdata"),
+						},
+					},
+				},
+			},
+		},
+	}
+	assert.NoError(t, lgr1.(*kvLedger).commitToPvtAndBlockStore(dataAtCrash))
+	testVerifyPvtData(t, lgr1, blockNumAtCrash, expectedPvtData)
+	bcInfo, err = lgr1.GetBlockchainInfo()
+	assert.NoError(t, err)
+	assert.Equal(t, blockNumAtCrash+1, bcInfo.Height)
+
+	isPvtStoreAhead, err = lgr1.(*kvLedger).isPvtDataStoreAheadOfBlockStore()
+	assert.NoError(t, err)
+	assert.False(t, isPvtStoreAhead)
+}
+
+func testVerifyPvtData(t *testing.T, ledger ledger.PeerLedger, blockNum uint64, expectedPvtData lgr.TxPvtDataMap) {
+	pvtdata, err := ledger.GetPvtDataByNum(blockNum, nil)
+	assert.NoError(t, err)
+	constructed := constructPvtdataMap(pvtdata)
+	assert.Equal(t, len(expectedPvtData), len(constructed))
+	for k, v := range expectedPvtData {
+		ov, ok := constructed[k]
+		assert.True(t, ok)
+		assert.Equal(t, v.SeqInBlock, ov.SeqInBlock)
+		assert.True(t, proto.Equal(v.WriteSet, ov.WriteSet))
+	}
+}
+
+func TestPvtStoreAheadOfBlockStore(t *testing.T) {
+	conf, cleanup := testConfig(t)
+	defer cleanup()
+	provider := testutilNewProvider(conf, t, true)
+	defer provider.Close()
+
+	ledgerID := "testLedger"
+	bg, gb := testutil.NewBlockGenerator(t, ledgerID, false)
+	gbHash := protoutil.BlockHeaderHash(gb.Header)
+	lgr, err := provider.Create(gb)
+	assert.NoError(t, err)
+	defer lgr.Close()
+
+	bcInfo, _ := lgr.GetBlockchainInfo()
+	assert.Equal(t, &common.BlockchainInfo{
+		Height: 1, CurrentBlockHash: gbHash, PreviousBlockHash: nil,
+	}, bcInfo)
+
+	// when both stores contain genesis block only, isPvtstoreAheadOfBlockstore should be false
+	kvlgr := lgr.(*kvLedger)
+	isPvtStoreAhead, err := kvlgr.isPvtDataStoreAheadOfBlockStore()
+	assert.NoError(t, err)
+	assert.False(t, isPvtStoreAhead)
+
+	sampleData := sampleDataWithPvtdataForSelectiveTx(t, bg)
+	for _, d := range sampleData[0:9] { // commit block number 0 to 8
+		assert.NoError(t, kvlgr.commitToPvtAndBlockStore(d))
+	}
+
+	isPvtStoreAhead, err = kvlgr.isPvtDataStoreAheadOfBlockStore()
+	assert.NoError(t, err)
+	assert.False(t, isPvtStoreAhead)
+
+	// close and reopen. It mimics peer crash.
```

Comment:
I removed the `mimics peer crash`.

`TestPvtStoreAheadOfBlockStore()` ensures that the `isPvtStoreAheadOfBlockStore` is set correctly. It can be set due to two scenarios: The block is written to the pvtdatastore and the peer fails before committing the block to blockstore. When the peer restarts, it should have `isPvtStoreAheadOfBlockStore` as true. Another scenario is related to rollback. Long ago we decided to have rollback related tests in the `kvledger/tests`. I am not understanding the problem with this test. If I miss something, please let me know.

Reply:
What I meant in my comment is the following....
Earlier, these tests were in a separate package and the scope of that package was clear, i.e., maintaining consistency between ledger storage and pvt data store (exactly, what you have described above). So, a meaning of a crash mimic is clear once you are in that package.
Now, we chose to move that code here in the kvledger package. The kvldegder package deals with much more coordination between many components. mimic crash could mean that "is this test mimicking a crash after writing to both blockstore and pvtdata store, but short of committing to statedb?".
That's what I meant that "TestPvtStoreAheadOfBlockStore" name does not convey what is the state of other stores and hence I made the comment that - "The tests that are moved here still have the flavor of testing of commits between the two store only. After some time, we may forget the scope of these tests and I guess that it would make sense to either have them in a separate test file or renaming the tests to makes the scope clear." But this is not something that should hold this PR from merging. This can be fixed later in a separate PR as well.